### ISSUE TITLE
Area rendering glitches with interactive filters

### ISSUE THREAD

**Issue by arvind** [vega/vega#930]
When rapidly dragging out a brush (left-to-right only), or when reversing the pan direction, the `detail` area chart exhibits a number of rendering artifacts. 

Perhaps something's going wrong with the ordering of input data tuples? But there's a `collect` transform in both datasets, and cursory inspection of `data_1` with a glitched output indicates that the tuples are indeed in descending order by date. 

![areaglitches](https://user-images.githubusercontent.com/42262/28292231-079bdb84-6b03-11e7-9118-e05f5a4d480e.gif)

```json
{
  "$schema": "https://vega.github.io/schema/vega/v3.0.json",
  "autosize": "pad",
  "padding": 5,
  "data": [
    {"name": "brush_store"},
    {
      "name": "source_0",
      "url": "data/sp500.csv",
      "format": {
        "type": "csv",
        "parse": {"date": "date","price": "number"}
      }
    },
    {
      "name": "data_0",
      "source": "source_0",
      "transform": [
        {
          "type": "collect",
          "sort": {"field": "date","order": "descending"}
        }
      ]
    },
    {
      "name": "data_1",
      "source": "source_0",
      "transform": [
        {
          "type": "filter",
          "expr": "vlInterval(\"brush_store\", \"concat_0_\", datum, \"union\", \"all\")"
        },
        {
          "type": "collect",
          "sort": {"field": "date","order": "descending"}
        }
      ]
    }
  ],
  "signals": [
    {"name": "width","update": "480"},
    {"name": "concat_0_height","update": "200"},
    {"name": "concat_1_height","update": "60"},
    {
      "name": "unit",
      "value": {},
      "on": [
        {
          "events": "mousemove",
          "update": "group()._id ? group() : unit"
        }
      ]
    }
  ],
  "layout": {
    "padding": {"row": 10,"column": 10},
    "offset": 10,
    "columns": 1,
    "bounds": "full",
    "align": "all"
  },
  "marks": [
    {
      "type": "group",
      "name": "concat_0_group",
      "encode": {
        "update": {
          "width": {"signal": "width"},
          "height": {"signal": "concat_0_height"},
          "fill": {"value": "transparent"}
        }
      },
      "marks": [
        {
          "name": "concat_0_marks",
          "type": "area",
          "from": {"data": "data_1"},
          "encode": {
            "update": {
              "x": {"scale": "concat_0_x","field": "date"},
              "y": {"scale": "concat_0_y","field": "price"},
              "y2": {"scale": "concat_0_y","value": 0},
              "fill": {"value": "#4c78a8"},
              "orient": {"value": "vertical"}
            }
          }
        }
      ],
      "scales": [
        {
          "name": "concat_0_x",
          "type": "time",
          "domain": {"data": "data_1","field": "date"},
          "range": [0,480]
        },
        {
          "name": "concat_0_y",
          "type": "linear",
          "domain": {"data": "data_1","field": "price"},
          "range": [200,0]
        }
      ]
    },
    {
      "type": "group",
      "name": "concat_1_group",
      "encode": {
        "update": {
          "width": {"signal": "width"},
          "height": {"signal": "concat_1_height"},
          "fill": {"value": "transparent"}
        }
      },
      "signals": [
        {
          "name": "brush_x",
          "value": [],
          "on": [
            {
              "events": {
                "source": "scope",
                "type": "mousedown",
                "filter": [
                  "!event.item || event.item.mark.name !== \"brush_brush\""
                ]
              },
              "update": "[x(unit), x(unit)]"
            },
            {
              "events": {
                "source": "window",
                "type": "mousemove",
                "consume": true,
                "between": [
                  {
                    "source": "scope",
                    "type": "mousedown",
                    "filter": [
                      "!event.item || event.item.mark.name !== \"brush_brush\""
                    ]
                  },
                  {"source": "window","type": "mouseup"}
                ]
              },
              "update": "[brush_x[0], clamp(x(unit), 0, width)]"
            },
            {
              "events": {"signal": "brush_translate_delta"},
              "update": "clampRange(panLinear(brush_translate_anchor.extent_x, brush_translate_delta.x / span(brush_translate_anchor.extent_x)), 0, width)"
            }
          ]
        },
        {
          "name": "brush_date",
          "on": [
            {
              "events": {"signal": "brush_x"},
              "update": "invert(\"concat_1_x\", brush_x)"
            }
          ]
        },
        {
          "name": "brush_tuple",
          "update": "{unit: \"concat_1_\", intervals: [{encoding: \"x\", field: \"date\", extent: brush_date}]}"
        },
        {
          "name": "brush_translate_anchor",
          "value": {},
          "on": [
            {
              "events": [
                {
                  "source": "scope",
                  "type": "mousedown",
                  "markname": "brush_brush"
                }
              ],
              "update": "{x: x(unit), y: y(unit), extent_x: slice(brush_x)}"
            }
          ]
        },
        {
          "name": "brush_translate_delta",
          "value": {},
          "on": [
            {
              "events": [
                {
                  "source": "window",
                  "type": "mousemove",
                  "consume": true,
                  "between": [
                    {
                      "source": "scope",
                      "type": "mousedown",
                      "markname": "brush_brush"
                    },
                    {"source": "window","type": "mouseup"}
                  ]
                }
              ],
              "update": "{x: brush_translate_anchor.x - x(unit), y: brush_translate_anchor.y - y(unit)}"
            }
          ]
        },
        {
          "name": "brush_modify",
          "on": [
            {
              "events": {"signal": "brush_tuple"},
              "update": "modify(\"brush_store\", brush_tuple, true)"
            }
          ]
        }
      ],
      "marks": [
        {
          "name": "concat_1_marks",
          "type": "area",
          "from": {"data": "data_0"},
          "encode": {
            "update": {
              "x": {"scale": "concat_1_x","field": "date"},
              "y": {"scale": "concat_1_y","field": "price"},
              "y2": {"scale": "concat_1_y","value": 0},
              "fill": {"value": "#4c78a8"},
              "orient": {"value": "vertical"}
            }
          }
        },
        {
          "name": "brush_brush",
          "type": "rect",
          "encode": {
            "enter": {
              "fill": {"value": "#333"},
              "fillOpacity": {"value": 0.125},
              "stroke": {"value": "white"}
            },
            "update": {
              "x": {"signal": "brush_x[0]"},
              "x2": {"signal": "brush_x[1]"},
              "y": {"value": 0},
              "y2": {"field": {"group": "height"}}
            }
          }
        }
      ],
      "scales": [
        {
          "name": "concat_1_x",
          "type": "time",
          "domain": {"data": "data_0","field": "date"},
          "range": [0,480]
        },
        {
          "name": "concat_1_y",
          "type": "linear",
          "domain": {"data": "data_0","field": "price"},
          "range": [60,0]
        }
      ]
    }
  ]
}
```

**Comment by jheer** [comment]
This is not a rendering glitch. If you check the scenegraph content you'll see that it is actually rendering correctly. The problem is that the scenegraph item order is getting jumbled by the filtering, which results in selectively removing and adding scenegraph items. Even though the backing _data store_ is sorted, the data join with the _scenegraph_ does not automatically preserve ordering upon intermittent future adds. To enforce a sort order among the scene items, you can use the mark-level `sort` property.

In the spec above you can remove the `collect` transforms and add the following line to the upper area mark definition: `"sort": {"field": "x"},`.

**Comment by jheer** [comment]
FYI, after fixing this spec, I discovered an unrelated bug in the SVGRenderer when removing a single-point area or line mark. A fix to vega-scenegraph was released in v2.0.2.
